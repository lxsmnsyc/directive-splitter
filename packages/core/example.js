import { compile } from 'dismantle';

const code = `
import { server$ } from 'my-example';

const foo = 'foo';
const bar = 'bar';

function log(prefix, suffix) {
  console.log(prefix, suffix);
}

const logPrefix = server$((prefix) => {
  if (prefix === 'foo') {
    log(prefix, foo);
  } else {
    log(prefix, bar);
  }
});
`;

const result = await compile(code, '/path/to/example.ts', {
  key: 'example',
  mode: 'server',
  env: 'development',
  directives: [
    {
      value: 'use server',
      import: {
        kind: 'named',
        name: '$$server',
        source: 'my-example',
      },
    },
  ],
  functions: [
    {
      source: {
        kind: 'named',
        name: 'server$',
        source: 'my-example',
      },
      target: {
        kind: 'named',
        name: 'server$',
        source: 'my-example/server',
      },
    },
  ],
});

console.log('ENTRY');
console.log(result.code);
console.log(result.map);

console.log('\nVIRTUALS');
for (const [name, content] of result.files) {
  console.log('FILE', name);
  console.log(content);
}

console.log('\nENTRIES');
console.log(result.entries);

console.log('\nROOTs');
console.log(result.roots);
