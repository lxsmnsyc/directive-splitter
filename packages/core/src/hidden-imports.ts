import type { ImportDefinition } from './types';

export const HIDDEN_FUNC: ImportDefinition = {
  kind: 'named',
  source: 'dismantle/runtime',
  name: '$$func',
};

export const HIDDEN_GENERATOR: ImportDefinition = {
  kind: 'named',
  source: 'dismantle/runtime',
  name: '$$gen',
};
