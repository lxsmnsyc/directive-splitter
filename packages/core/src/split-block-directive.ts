import type * as babel from '@babel/core';
import * as t from '@babel/types';
import { HIDDEN_FUNC, HIDDEN_GENERATOR } from './hidden-imports';
import type { ExtractedBindings } from './split';
import {
  BREAK_KEY,
  CONTINUE_KEY,
  NO_HALT_KEY,
  RETURN_KEY,
  THROW_KEY,
  YIELD_KEY,
  createEntryFile,
  createRootFile,
  extractBindings,
  getGeneratorReplacementForBlock,
} from './split';
import type { BlockDirectiveDefinition, StateContext } from './types';
import { generateUniqueName } from './utils/generate-unique-name';
import getForeignBindings from './utils/get-foreign-bindings';
import { getImportIdentifier } from './utils/get-import-identifier';

interface HaltingBlockResult {
  breaks: string[];
  breakCount: number;
  continues: string[];
  continueCount: number;
  hasReturn: boolean;
  hasYield: boolean;
}

function transformBlockContent(
  path: babel.NodePath<t.BlockStatement>,
  mutations: t.Identifier[],
): HaltingBlockResult {
  const target =
    path.scope.getFunctionParent() || path.scope.getProgramParent();

  const breaks: string[] = [];
  let breakCount = 0;
  const continues: string[] = [];
  let continueCount = 0;
  let hasReturn = false;
  let hasYield = false;

  const applyMutations = mutations.length
    ? generateUniqueName(path, 'mutate')
    : undefined;

  // Transform the control flow statements
  path.traverse({
    BreakStatement(child) {
      const parent =
        child.scope.getFunctionParent() || child.scope.getProgramParent();
      if (parent === target) {
        const replacement: t.Expression[] = [BREAK_KEY];
        breakCount++;
        if (child.node.label) {
          const targetName = child.node.label.name;
          breaks.push(targetName);
          replacement.push(t.stringLiteral(targetName));
        } else {
          replacement.push(t.nullLiteral());
        }
        if (applyMutations) {
          replacement.push(t.callExpression(applyMutations, []));
        }
        child.replaceWith(t.returnStatement(t.arrayExpression(replacement)));
        child.skip();
      }
    },
    ContinueStatement(child) {
      const parent =
        child.scope.getFunctionParent() || child.scope.getProgramParent();
      if (parent === target) {
        const replacement: t.Expression[] = [CONTINUE_KEY];
        continueCount++;
        if (child.node.label) {
          const targetName = child.node.label.name;
          continues.push(targetName);
          replacement.push(t.stringLiteral(targetName));
        } else {
          replacement.push(t.nullLiteral());
        }
        if (applyMutations) {
          replacement.push(t.callExpression(applyMutations, []));
        }
        child.replaceWith(t.returnStatement(t.arrayExpression(replacement)));
        child.skip();
      }
    },
    ReturnStatement(child) {
      const parent =
        child.scope.getFunctionParent() || child.scope.getProgramParent();
      if (parent === target) {
        hasReturn = true;
        const arg = child.get('argument');
        arg.replaceWith(
          t.arrayExpression([
            RETURN_KEY,
            arg.node ? arg.node : t.nullLiteral(),
            applyMutations
              ? t.callExpression(applyMutations, [])
              : t.nullLiteral(),
          ]),
        );
      }
    },
    YieldExpression(child) {
      const parent =
        child.scope.getFunctionParent() || child.scope.getProgramParent();
      if (parent === target) {
        hasYield = true;
        if (child.node.delegate) {
          // TODO
        } else {
          const arg = child.get('argument');
          arg.replaceWith(
            t.arrayExpression([
              YIELD_KEY,
              arg.node ? arg.node : t.nullLiteral(),
              applyMutations
                ? t.callExpression(applyMutations, [])
                : t.nullLiteral(),
            ]),
          );
        }
      }
    },
  });

  const error = generateUniqueName(path, 'error');

  const throwResult: t.Expression[] = [THROW_KEY, error];
  const haltResult: t.Expression[] = [NO_HALT_KEY];

  if (applyMutations) {
    throwResult.push(t.callExpression(applyMutations, []));
    haltResult.push(t.nullLiteral());
    haltResult.push(t.callExpression(applyMutations, []));
  }

  const statements: t.Statement[] = [
    t.tryStatement(
      t.blockStatement(path.node.body),
      t.catchClause(
        error,
        t.blockStatement([t.returnStatement(t.arrayExpression(throwResult))]),
      ),
    ),
    t.returnStatement(t.arrayExpression(haltResult)),
  ];

  if (applyMutations) {
    statements.unshift(
      t.variableDeclaration('const', [
        t.variableDeclarator(
          applyMutations,
          t.arrowFunctionExpression(
            [],
            t.objectExpression(
              mutations.map(item => t.objectProperty(item, item, false, true)),
            ),
          ),
        ),
      ]),
    );
  }

  path.node.body = statements;
  return { breaks, continues, hasReturn, hasYield, breakCount, continueCount };
}

/**
 * This generates a chain of if-statements that checks the
 * received server return (which is transformed from the original block's
 * server statement)
 * Each if-statement matches an specific label, assuming that the original
 * break statement is a labeled break statement.
 * Otherwise, the output code is either a normal break statement or none.
 */
function getBreakCheck(
  returnType: t.Identifier,
  returnResult: t.Identifier,
  breakCount: number,
  breaks: string[],
  check: t.Statement | undefined,
): t.Statement | undefined {
  let current: t.Statement | undefined;
  if (breakCount !== breaks.length) {
    current = t.blockStatement([t.breakStatement()]);
  }
  for (let i = 0, len = breaks.length; i < len; i++) {
    const target = breaks[i];
    current = t.blockStatement([
      t.ifStatement(
        t.binaryExpression('===', returnResult, t.stringLiteral(target)),
        t.blockStatement([t.breakStatement(t.identifier(target))]),
        current,
      ),
    ]);
  }
  if (current) {
    return t.ifStatement(
      t.binaryExpression('===', returnType, BREAK_KEY),
      current,
      check,
    );
  }
  return check;
}

/**
 * This generates a chain of if-statements that checks the
 * received server return (which is transformed from the original block's
 * server statement)
 * Each if-statement matches an specific label, assuming that the original
 * continue statement is a labeled continue statement.
 * Otherwise, the output code is either a normal continue statement or none.
 */
function getContinueCheck(
  returnType: t.Identifier,
  returnResult: t.Identifier,
  continueCount: number,
  continues: string[],
  check: t.Statement | undefined,
): t.Statement | undefined {
  let current: t.Statement | undefined;
  if (continueCount !== continues.length) {
    current = t.blockStatement([t.continueStatement()]);
  }
  for (let i = 0, len = continues.length; i < len; i++) {
    const target = continues[i];
    current = t.blockStatement([
      t.ifStatement(
        t.binaryExpression('===', returnResult, t.stringLiteral(target)),
        t.blockStatement([t.continueStatement(t.identifier(target))]),
        current,
      ),
    ]);
  }
  if (current) {
    return t.ifStatement(
      t.binaryExpression('===', returnType, CONTINUE_KEY),
      current,
      check,
    );
  }
  return check;
}

function getBlockDirectiveReplacement(
  ctx: StateContext,
  path: babel.NodePath<t.BlockStatement>,
  entryFile: string,
  bindings: ExtractedBindings,
  halting: HaltingBlockResult,
) {
  // Move to the replacement for the server block,
  // declare the type and result based from transformBlockContent
  const returnType = generateUniqueName(path, 'type');
  const returnResult = generateUniqueName(path, 'result');
  const returnMutations = generateUniqueName(path, 'mutations');
  let check: t.Statement | undefined;
  // If the block has a return, we need to make sure that the
  // replacement does too.
  if (halting.hasReturn) {
    check = t.ifStatement(
      t.binaryExpression('===', returnType, RETURN_KEY),
      t.blockStatement([t.returnStatement(returnResult)]),
      check,
    );
  }
  // If the block has a break, we also do it.
  if (halting.breakCount > 0) {
    check = getBreakCheck(
      returnType,
      returnResult,
      halting.breakCount,
      halting.breaks,
      check,
    );
  }
  // If the block has a continue, we also do it.
  if (halting.continueCount > 0) {
    check = getContinueCheck(
      returnType,
      returnResult,
      halting.continueCount,
      halting.continues,
      check,
    );
  }
  const replacement: t.Statement[] = [];
  if (halting.hasYield) {
    const blockID = generateUniqueName(path, 'block');
    replacement.push(
      t.variableDeclaration('const', [
        t.variableDeclarator(
          blockID,
          t.callExpression(
            getImportIdentifier(ctx, path, {
              ...HIDDEN_GENERATOR,
              source: ctx.options.runtime,
            }),
            [
              t.memberExpression(
                t.awaitExpression(
                  t.importExpression(t.stringLiteral(entryFile)),
                ),
                t.identifier('default'),
              ),
              bindings.mutations.length
                ? t.arrowFunctionExpression(
                    [returnMutations],
                    t.assignmentExpression(
                      '=',
                      t.objectPattern(
                        bindings.mutations.map(item =>
                          t.objectProperty(item, item, false, true),
                        ),
                      ),
                      returnMutations,
                    ),
                  )
                : t.nullLiteral(),
            ],
          ),
        ),
      ]),
    );
    const [reps, step] = getGeneratorReplacementForBlock(
      path,
      blockID,
      bindings.locals,
    );
    for (let i = 0, len = reps.length; i < len; i++) {
      replacement.push(reps[i]);
    }
    replacement.push(
      t.variableDeclaration('const', [
        t.variableDeclarator(
          t.arrayPattern([returnType, returnResult]),
          t.memberExpression(step, t.identifier('value')),
        ),
      ]),
    );
  } else {
    replacement.push(
      t.variableDeclaration('const', [
        t.variableDeclarator(
          t.arrayPattern([returnType, returnResult]),
          t.awaitExpression(
            t.callExpression(
              t.callExpression(
                getImportIdentifier(ctx, path, {
                  ...HIDDEN_FUNC,
                  source: ctx.options.runtime,
                }),
                [
                  t.memberExpression(
                    t.awaitExpression(
                      t.importExpression(t.stringLiteral(entryFile)),
                    ),
                    t.identifier('default'),
                  ),
                  bindings.mutations.length
                    ? t.arrowFunctionExpression(
                        [returnMutations],
                        t.assignmentExpression(
                          '=',
                          t.objectPattern(
                            bindings.mutations.map(item =>
                              t.objectProperty(item, item, false, true),
                            ),
                          ),
                          returnMutations,
                        ),
                      )
                    : t.nullLiteral(),
                ],
              ),
              bindings.locals,
            ),
          ),
        ),
      ]),
    );
  }
  if (check) {
    replacement.push(check);
  }

  return t.blockStatement(replacement);
}

function replaceBlockDirective(
  ctx: StateContext,
  path: babel.NodePath<t.BlockStatement>,
  directive: BlockDirectiveDefinition,
  bindings: ExtractedBindings,
): void {
  // Transform all control statements
  const halting = transformBlockContent(path, bindings.mutations);
  const entryFile = createEntryFile(
    ctx,
    path,
    ctx.options.mode === 'server' || directive.isomorphic
      ? createRootFile(
          ctx,
          bindings,
          t.functionExpression(
            undefined,
            bindings.locals,
            t.blockStatement(path.node.body),
            halting.hasYield,
            true,
          ),
        )
      : undefined,
    directive.target,
    directive.idPrefix,
  );
  path.replaceWith(
    getBlockDirectiveReplacement(ctx, path, entryFile, bindings, halting),
  );
}

export function splitBlockDirective(
  ctx: StateContext,
  path: babel.NodePath<t.BlockStatement>,
  directive: BlockDirectiveDefinition,
): void {
  replaceBlockDirective(
    ctx,
    path,
    directive,
    extractBindings(
      ctx,
      path,
      getForeignBindings(path, 'block'),
      directive.pure,
    ),
  );
}
