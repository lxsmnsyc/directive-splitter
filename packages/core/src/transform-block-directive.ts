import type * as babel from '@babel/core';
import type * as t from '@babel/types';
import { splitBlockDirective } from './split-block-directive';
import type { BlockDirectiveDefinition, StateContext } from './types';
import {
  cleanBlockForDirectives,
  cleanBlockForFauxDirectives,
  getDefinitionFromDirectives,
  getDefinitionFromFauxDirectives,
} from './utils/directive-check';

function getBlockDirectiveDefinition(
  ctx: StateContext,
  path: babel.NodePath<t.BlockStatement>,
): BlockDirectiveDefinition | undefined {
  const definition = getDefinitionFromDirectives(ctx, 'block-directive', path);
  if (definition) {
    cleanBlockForDirectives(path, definition);
    return definition;
  }
  const fauxDefinition = getDefinitionFromFauxDirectives(
    ctx,
    'block-directive',
    path,
  );
  if (fauxDefinition) {
    cleanBlockForFauxDirectives(path, fauxDefinition);
    return fauxDefinition;
  }
  return undefined;
}

export function transformBlockDirective(
  ctx: StateContext,
  path: babel.NodePath<t.BlockStatement>,
): void {
  const definition = getBlockDirectiveDefinition(ctx, path);
  if (definition) {
    splitBlockDirective(ctx, path, definition);
  }
}
