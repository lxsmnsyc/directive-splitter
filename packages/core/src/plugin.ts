import type * as babel from '@babel/core';
import { transformCall } from './transform-call';
import { transformBlockDirective } from './transform-block-directive';
import type { StateContext } from './types';
import { registerImportSpecifiers } from './utils/register-import-specifiers';

function bubbleFunctionDeclaration(): void {

}

interface State extends babel.PluginPass {
  opts: StateContext;
}

export function plugin(): babel.PluginObj<State> {
  return {
    name: 'dismantle',
    visitor: {
      Program(path, ctx) {
        registerImportSpecifiers(ctx.opts, path);
      },
      ArrowFunctionExpression(path, ctx) {

      },
      FunctionExpression(path, ctx) {

      },
      FunctionDeclaration(path, ctx) {

      },
      BlockStatement(path, ctx) {
        transformBlockDirective(ctx.opts, path);
      },
      CallExpression(path, ctx) {
        transformCall(ctx.opts, path);
      },
    },
  };
}
