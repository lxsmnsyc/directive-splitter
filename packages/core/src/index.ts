import * as babel from '@babel/core';
import path from 'path';
import type { Options, StateContext } from './types';
import assert from './utils/assert';
import { plugin } from './plugin';
import xxHash32 from './utils/xxhash32';

export interface Output {
  code: babel.BabelFileResult['code'];
  map: babel.BabelFileResult['map'];
  files: Map<string, string>;
}

export {
  NamedImportDefinition,
  DefaultImportDefinition,
  DirectiveDefinition,
  ImportDefinition,
  Options,
} from './types';

export async function compile(
  code: string,
  id: string,
  options: Options,
): Promise<Output> {
  const parsedPath = path.parse(id);

  const ctx: StateContext = {
    path: parsedPath,
    imports: new Map(),
    virtual: {
      files: new Map(),
      count: 0,
    },
    options,
    bindings: new Map(),
    blocks: {
      hash: xxHash32(id).toString(16),
      count: 0,
    },
  };

  const plugins: babel.ParserOptions['plugins'] = ['jsx'];

  if (/\.[mc]?tsx?$/i.test(id)) {
    plugins.push('typescript');
  }

  const result = await babel.transformAsync(code, {
    plugins: [[plugin, ctx]],
    parserOpts: {
      plugins,
    },
    filename: parsedPath.base,
    ast: false,
    sourceFileName: id,
    sourceMaps: true,
    configFile: false,
    babelrc: false,
  });

  assert(result, 'invariant');

  const files = new Map();

  for (const [key, value] of ctx.virtual.files) {
    files.set(path.join(parsedPath.dir, key), value);
  }

  return {
    code: result.code,
    map: result.map,
    files,
  };
}
