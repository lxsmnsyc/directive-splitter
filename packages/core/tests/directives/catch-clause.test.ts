import { describe, it, expect } from 'vitest';
import * as compiler from '../../src';
import { ID, CLIENT, SERVER } from './example';

describe('CatchClause', () => {
  describe('client', () => {
    it('should transform valid server try statements', async () => {
      const code = `
      try {
        doStuff();
      } catch (err) {
        'use server';
        await report(err);
      }
      `;
      expect(await compiler.compile(code, ID, CLIENT)).toMatchSnapshot();
    });
    it('should skip server try statements in non-async functions', async () => {
      const code = `
      const example = () => {
        try {
          doStuff();
        } catch (err) {
          'use server';
          report(err);
        }
      };
      `;
      expect(await compiler.compile(code, ID, CLIENT)).toMatchSnapshot();
    });
    it('should transform valid server functions with scope', async () => {
      const code = `
      async function foo() {
        const value = 'foo bar';
        try {
          doStuff();
        } catch (err) {
          'use server';
          await report(value, err);
        }
      }
      `;
      expect(await compiler.compile(code, ID, CLIENT)).toMatchSnapshot();
    });
    it('should skip top-level values for scope', async () => {
      const code = `
      const value = 'foo bar';
      try {
        doStuff();
      } catch (err) {
        'use server';
        await report(value, err);
      }
      `;
      expect(await compiler.compile(code, ID, CLIENT)).toMatchSnapshot();
    });
  });
  describe('server', () => {
    it('should transform valid server try statements', async () => {
      const code = `
      try {
        doStuff();
      } catch (err) {
        'use server';
        await report(err);
      }
      `;
      expect(await compiler.compile(code, ID, SERVER)).toMatchSnapshot();
    });
    it('should skip server try statements in non-async functions', async () => {
      const code = `
      const example = () => {
        try {
          doStuff();
        } catch (err) {
          'use server';
          report(err);
        }
      };
      `;
      expect(await compiler.compile(code, ID, SERVER)).toMatchSnapshot();
    });
    it('should transform valid server functions with scope', async () => {
      const code = `
      async function foo() {
        const value = 'foo bar';
        try {
          doStuff();
        } catch (err) {
          'use server';
          await report(err);
        }
      }
      `;
      expect(await compiler.compile(code, ID, SERVER)).toMatchSnapshot();
    });
    it('should skip top-level values for scope', async () => {
      const code = `
      const value = 'foo bar';
      try {
        doStuff();
      } catch (err) {
        'use server';
        await report(err);
      }
      `;
      expect(await compiler.compile(code, ID, SERVER)).toMatchSnapshot();
    });
  });
});
