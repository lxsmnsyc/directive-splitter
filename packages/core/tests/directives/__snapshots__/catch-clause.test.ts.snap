// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CatchClause > client > should skip server try statements in non-async functions 1`] = `
{
  "code": "const example = () => {
  try {
    doStuff();
  } catch (err) {
    'use server';
    report(err);
  }
};",
  "entries": [],
  "files": Map {},
  "map": {
    "file": undefined,
    "mappings": "AACM,MAAMA,OAAO,GAAGA,CAAA,KAAM;EACpB,IAAI;IACFC,OAAO,CAAC,CAAC;EACX,CAAC,CAAC,OAAOC,GAAG,EAAE;IACZ,YAAY;IACZC,MAAM,CAACD,GAAG,CAAC;EACb;AACF,CAAC",
    "names": [
      "example",
      "doStuff",
      "err",
      "report",
    ],
    "sourceRoot": undefined,
    "sources": [
      "/path/to/example.ts",
    ],
    "sourcesContent": [
      "
      const example = () => {
        try {
          doStuff();
        } catch (err) {
          'use server';
          report(err);
        }
      };
      ",
    ],
    "version": 3,
  },
  "roots": [],
}
`;

exports[`CatchClause > client > should skip top-level values for scope 1`] = `
{
  "code": "import { value } from "./example.ts?my-example=0.ts";
try {
  doStuff();
} catch (err) {
  const _block = (await import("./example.ts?my-example=2.ts")).default;
  const [_type, _result, _mutations] = await _block();
  if (_type === 4) {
    throw _result;
  }
}",
  "entries": [
    "/path/to/example.ts?my-example=2.ts",
  ],
  "files": Map {
    "/path/to/example.ts?my-example=0.ts" => "export const value = 'foo bar';",
    "/path/to/example.ts?my-example=1.ts" => "export default (async function () {
  try {
    await report(value, err);
  } catch (_error) {
    return [4, _error];
  }
  return [3];
});",
    "/path/to/example.ts?my-example=2.ts" => "import { server as _entry } from "my-example/client";
export default _entry("26ecc3a5-0-anonymous");",
  },
  "map": {
    "file": undefined,
    "mappings": ";AAEM,IAAI;EACFA,OAAO,CAAC,CAAC;AACX,CAAC,CAAC,OAAOC,GAAG;EAAA,MAAAC,MAAA,kDAAAC,OAAA;EAAA,OAAAC,KAAA,EAAAC,OAAA,EAAAC,UAAA,UAAAJ,MAAA;EAAA,IAAAE,KAAA;IAAA,MAAAC,OAAA;EAAA;AAAA",
    "names": [
      "doStuff",
      "err",
      "_block",
      "default",
      "_type",
      "_result",
      "_mutations",
    ],
    "sourceRoot": undefined,
    "sources": [
      "/path/to/example.ts",
    ],
    "sourcesContent": [
      "
      const value = 'foo bar';
      try {
        doStuff();
      } catch (err) {
        'use server';
        await report(value, err);
      }
      ",
    ],
    "version": 3,
  },
  "roots": [
    "/path/to/example.ts?my-example=1.ts",
  ],
}
`;

exports[`CatchClause > client > should transform valid server functions with scope 1`] = `
{
  "code": "async function foo() {
  const value = 'foo bar';
  try {
    doStuff();
  } catch (err) {
    const _block = (await import("./example.ts?my-example=1.ts")).default;
    const [_type, _result, _mutations] = await _block(value);
    if (_type === 4) {
      throw _result;
    }
  }
}",
  "entries": [
    "/path/to/example.ts?my-example=1.ts",
  ],
  "files": Map {
    "/path/to/example.ts?my-example=0.ts" => "export default (async function (value) {
  try {
    await report(value, err);
  } catch (_error) {
    return [4, _error];
  }
  return [3];
});",
    "/path/to/example.ts?my-example=1.ts" => "import { server as _entry } from "my-example/client";
export default _entry("26ecc3a5-0-foo");",
  },
  "map": {
    "file": undefined,
    "mappings": "AACM,eAAeA,GAAGA,CAAA,EAAG;EACnB,MAAMC,KAAK,GAAG,SAAS;EACvB,IAAI;IACFC,OAAO,CAAC,CAAC;EACX,CAAC,CAAC,OAAOC,GAAG;IAAA,MAAAC,MAAA,kDAAAC,OAAA;IAAA,OAAAC,KAAA,EAAAC,OAAA,EAAAC,UAAA,UAAAJ,MAAA,CAHNH,KAAK;IAAA,IAAAK,KAAA;MAAA,MAAAC,OAAA;IAAA;EAAA;AAOb",
    "names": [
      "foo",
      "value",
      "doStuff",
      "err",
      "_block",
      "default",
      "_type",
      "_result",
      "_mutations",
    ],
    "sourceRoot": undefined,
    "sources": [
      "/path/to/example.ts",
    ],
    "sourcesContent": [
      "
      async function foo() {
        const value = 'foo bar';
        try {
          doStuff();
        } catch (err) {
          'use server';
          await report(value, err);
        }
      }
      ",
    ],
    "version": 3,
  },
  "roots": [
    "/path/to/example.ts?my-example=0.ts",
  ],
}
`;

exports[`CatchClause > client > should transform valid server try statements 1`] = `
{
  "code": "try {
  doStuff();
} catch (err) {
  const _block = (await import("./example.ts?my-example=1.ts")).default;
  const [_type, _result, _mutations] = await _block();
  if (_type === 4) {
    throw _result;
  }
}",
  "entries": [
    "/path/to/example.ts?my-example=1.ts",
  ],
  "files": Map {
    "/path/to/example.ts?my-example=0.ts" => "export default (async function () {
  try {
    await report(err);
  } catch (_error) {
    return [4, _error];
  }
  return [3];
});",
    "/path/to/example.ts?my-example=1.ts" => "import { server as _entry } from "my-example/client";
export default _entry("26ecc3a5-0-anonymous");",
  },
  "map": {
    "file": undefined,
    "mappings": "AACM,IAAI;EACFA,OAAO,CAAC,CAAC;AACX,CAAC,CAAC,OAAOC,GAAG;EAAA,MAAAC,MAAA,kDAAAC,OAAA;EAAA,OAAAC,KAAA,EAAAC,OAAA,EAAAC,UAAA,UAAAJ,MAAA;EAAA,IAAAE,KAAA;IAAA,MAAAC,OAAA;EAAA;AAAA",
    "names": [
      "doStuff",
      "err",
      "_block",
      "default",
      "_type",
      "_result",
      "_mutations",
    ],
    "sourceRoot": undefined,
    "sources": [
      "/path/to/example.ts",
    ],
    "sourcesContent": [
      "
      try {
        doStuff();
      } catch (err) {
        'use server';
        await report(err);
      }
      ",
    ],
    "version": 3,
  },
  "roots": [
    "/path/to/example.ts?my-example=0.ts",
  ],
}
`;

exports[`CatchClause > server > should skip server try statements in non-async functions 1`] = `
{
  "code": "const example = () => {
  try {
    doStuff();
  } catch (err) {
    'use server';
    report(err);
  }
};",
  "entries": [],
  "files": Map {},
  "map": {
    "file": undefined,
    "mappings": "AACM,MAAMA,OAAO,GAAGA,CAAA,KAAM;EACpB,IAAI;IACFC,OAAO,CAAC,CAAC;EACX,CAAC,CAAC,OAAOC,GAAG,EAAE;IACZ,YAAY;IACZC,MAAM,CAACD,GAAG,CAAC;EACb;AACF,CAAC",
    "names": [
      "example",
      "doStuff",
      "err",
      "report",
    ],
    "sourceRoot": undefined,
    "sources": [
      "/path/to/example.ts",
    ],
    "sourcesContent": [
      "
      const example = () => {
        try {
          doStuff();
        } catch (err) {
          'use server';
          report(err);
        }
      };
      ",
    ],
    "version": 3,
  },
  "roots": [],
}
`;

exports[`CatchClause > server > should skip top-level values for scope 1`] = `
{
  "code": "const value = 'foo bar';
try {
  doStuff();
} catch (err) {
  const _block = (await import("./example.ts?my-example=1.ts")).default;
  const [_type, _result, _mutations] = await _block();
  if (_type === 4) {
    throw _result;
  }
}",
  "entries": [
    "/path/to/example.ts?my-example=1.ts",
  ],
  "files": Map {
    "/path/to/example.ts?my-example=0.ts" => "export default (async function () {
  try {
    await report(err);
  } catch (_error) {
    return [4, _error];
  }
  return [3];
});",
    "/path/to/example.ts?my-example=1.ts" => "import { server as _entry } from "my-example/server";
import _root from "./example.ts?my-example=0.ts";
export default _entry("26ecc3a5-0-anonymous", _root);",
  },
  "map": {
    "file": undefined,
    "mappings": "AACM,MAAMA,KAAK,GAAG,SAAS;AACvB,IAAI;EACFC,OAAO,CAAC,CAAC;AACX,CAAC,CAAC,OAAOC,GAAG;EAAA,MAAAC,MAAA,kDAAAC,OAAA;EAAA,OAAAC,KAAA,EAAAC,OAAA,EAAAC,UAAA,UAAAJ,MAAA;EAAA,IAAAE,KAAA;IAAA,MAAAC,OAAA;EAAA;AAAA",
    "names": [
      "value",
      "doStuff",
      "err",
      "_block",
      "default",
      "_type",
      "_result",
      "_mutations",
    ],
    "sourceRoot": undefined,
    "sources": [
      "/path/to/example.ts",
    ],
    "sourcesContent": [
      "
      const value = 'foo bar';
      try {
        doStuff();
      } catch (err) {
        'use server';
        await report(err);
      }
      ",
    ],
    "version": 3,
  },
  "roots": [
    "/path/to/example.ts?my-example=0.ts",
  ],
}
`;

exports[`CatchClause > server > should transform valid server functions with scope 1`] = `
{
  "code": "async function foo() {
  const value = 'foo bar';
  try {
    doStuff();
  } catch (err) {
    const _block = (await import("./example.ts?my-example=1.ts")).default;
    const [_type, _result, _mutations] = await _block();
    if (_type === 4) {
      throw _result;
    }
  }
}",
  "entries": [
    "/path/to/example.ts?my-example=1.ts",
  ],
  "files": Map {
    "/path/to/example.ts?my-example=0.ts" => "export default (async function () {
  try {
    await report(err);
  } catch (_error) {
    return [4, _error];
  }
  return [3];
});",
    "/path/to/example.ts?my-example=1.ts" => "import { server as _entry } from "my-example/server";
import _root from "./example.ts?my-example=0.ts";
export default _entry("26ecc3a5-0-foo", _root);",
  },
  "map": {
    "file": undefined,
    "mappings": "AACM,eAAeA,GAAGA,CAAA,EAAG;EACnB,MAAMC,KAAK,GAAG,SAAS;EACvB,IAAI;IACFC,OAAO,CAAC,CAAC;EACX,CAAC,CAAC,OAAOC,GAAG;IAAA,MAAAC,MAAA,kDAAAC,OAAA;IAAA,OAAAC,KAAA,EAAAC,OAAA,EAAAC,UAAA,UAAAJ,MAAA;IAAA,IAAAE,KAAA;MAAA,MAAAC,OAAA;IAAA;EAAA;AAId",
    "names": [
      "foo",
      "value",
      "doStuff",
      "err",
      "_block",
      "default",
      "_type",
      "_result",
      "_mutations",
    ],
    "sourceRoot": undefined,
    "sources": [
      "/path/to/example.ts",
    ],
    "sourcesContent": [
      "
      async function foo() {
        const value = 'foo bar';
        try {
          doStuff();
        } catch (err) {
          'use server';
          await report(err);
        }
      }
      ",
    ],
    "version": 3,
  },
  "roots": [
    "/path/to/example.ts?my-example=0.ts",
  ],
}
`;

exports[`CatchClause > server > should transform valid server try statements 1`] = `
{
  "code": "try {
  doStuff();
} catch (err) {
  const _block = (await import("./example.ts?my-example=1.ts")).default;
  const [_type, _result, _mutations] = await _block();
  if (_type === 4) {
    throw _result;
  }
}",
  "entries": [
    "/path/to/example.ts?my-example=1.ts",
  ],
  "files": Map {
    "/path/to/example.ts?my-example=0.ts" => "export default (async function () {
  try {
    await report(err);
  } catch (_error) {
    return [4, _error];
  }
  return [3];
});",
    "/path/to/example.ts?my-example=1.ts" => "import { server as _entry } from "my-example/server";
import _root from "./example.ts?my-example=0.ts";
export default _entry("26ecc3a5-0-anonymous", _root);",
  },
  "map": {
    "file": undefined,
    "mappings": "AACM,IAAI;EACFA,OAAO,CAAC,CAAC;AACX,CAAC,CAAC,OAAOC,GAAG;EAAA,MAAAC,MAAA,kDAAAC,OAAA;EAAA,OAAAC,KAAA,EAAAC,OAAA,EAAAC,UAAA,UAAAJ,MAAA;EAAA,IAAAE,KAAA;IAAA,MAAAC,OAAA;EAAA;AAAA",
    "names": [
      "doStuff",
      "err",
      "_block",
      "default",
      "_type",
      "_result",
      "_mutations",
    ],
    "sourceRoot": undefined,
    "sources": [
      "/path/to/example.ts",
    ],
    "sourcesContent": [
      "
      try {
        doStuff();
      } catch (err) {
        'use server';
        await report(err);
      }
      ",
    ],
    "version": 3,
  },
  "roots": [
    "/path/to/example.ts?my-example=0.ts",
  ],
}
`;
